Refactor lookup_type, fix non-function-like macro, error on stack arrays

gtker opened the pull request:
This PR:

* Adds an error for arrays on the stack.
* Makes a regular macro follow by a `(` token no longer an error since this broke the often used pattern of renaming functions with macros.
* Refactors `lookup_type` to `lookup_primitive_type` and `lookup_global_type`. These will be necessary for dealing with multi token types like `unsigned int` and `long long long long long long int`.

@stikonas @oriansj

stikonas reviewed (changes requested):
You lost part of the commit message in the second commit. Presumably because you started a line with `#`

gtker commented:
> You lost part of the commit message in the second commit. Presumably because you started a line with #

Ah whoops. Fixed.

I also added implicit `return 0` in the `main` function rather than just returning whatever was in the register at the time.
